src/python/dxpy/utils/config.py

```diff
@@ -0,0 +1,297 @@
+                    # (including, most commonly, because the sessions dir
+                    # doesn't exist)
+                    session_dirs = []
+                for session_dir in session_dirs:
+                    if not pid_exists(int(session_dir)):
+                        rmtree(os.path.join(sessions_dir, session_dir), ignore_errors=True)
+
+            parent_process = Process(os.getpid()).parent()
+            default_session_dir = os.path.join(sessions_dir, str(parent_process.pid))
+            while parent_process is not None and parent_process.pid != 0:
+                session_dir = os.path.join(sessions_dir, str(parent_process.pid))
+                if os.path.exists(session_dir):
+                    return session_dir
+                parent_process = parent_process.parent()
+            return default_session_dir
+        except (ImportError, IOError, AttributeError) as e:
+            warn(fill("Error while retrieving session configuration: " + format_exception(e)))
+        except Exception as e:
+            warn(fill("Unexpected error while retrieving session configuration: " + format_exception(e)))
+        return self._get_ppid_session_conf_dir(sessions_dir)
+
+    def _get_ppid_session_conf_dir(self, sessions_dir):
+        try:
+            return os.path.join(sessions_dir, str(os.getppid()))
+        except AttributeError:
+            pass # os.getppid is not available on Windows
+        except Exception as e:
+            warn(fill("Unexpected error while retrieving session configuration: " + format_exception(e)))
+        return os.path.join(sessions_dir, str(os.getpid()))
+
+    def _read_conf_dir(self, dirname):
+        try:
+            with open(os.path.join(dirname, "environment.json")) as fd:
+                env_vars = json.load(fd)
+        except Exception:
+            env_vars = {}
+
+        for standalone_var in self.STANDALONE_VAR_NAMES:
+            try:
+                with open(os.path.join(dirname, standalone_var)) as fd:
+                    env_vars[standalone_var] = fd.read()
+            except Exception:
+                pass
+        return env_vars
+
+    def __getitem__(self, item):
+        if item not in self.VAR_NAMES:
+            raise KeyError(item)
+        return environ[item]
+
+    def __setitem__(self, key, value):
+        if key not in self.VAR_NAMES:
+            raise KeyError(key)
+        if value is None:
+            value = self.defaults.get(key, "")
+        environ[key] = value
+        if key in self.CORE_VAR_NAMES:
+            self._sync_dxpy_state()
+
+    def __delitem__(self, key):
+        if key not in self.VAR_NAMES:
+            raise KeyError(key)
+        del environ[key]
+        if key in self.CORE_VAR_NAMES:
+            self._sync_dxpy_state()
+
+    def __iter__(self):
+        for item in self.VAR_NAMES:
+            if item in environ:
+                yield item
+
+    def __len__(self):
+        return len(self.VAR_NAMES)
```

Comment:
The length reported here ought to be consistent with the number of items returned by the iterator.